### PR TITLE
[JENKINS-29936] when removing an item use ACL.SYSTEM.

### PR DESCRIPTION
**Pull request by jtnord**
[JENKINS-29936](https://issues.jenkins-ci.org/browse/JENKINS-29936)

The OldDataMonitor should be using ACL.system not the ACL of the calling
thread - this also avoids the deadlock when an authorization strategy is
being saved (locking the auth strategy) which will call into the ODM at
the same point the ODM is being called an a Run has been saved (which will
cause a lookup of the job which will do a permissions check).

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/diagnosis/OldDataMonitor.java [hunk 1]
```diff
@@ -102,12 +110,21 @@ public boolean isActivated() {
...
+            SecurityContext oldContext = ACL.impersonate(ACL.SYSTEM);
+            try {
+                synchronized (odm) {
+                    odm.data.remove(referTo(obj));
+                    if (isDelete && obj instanceof Job<?,?>)
+                        for (Run r : ((Job<?,?>)obj).getBuilds())
+                            odm.data.remove(referTo(r));
+                }
+            } 
+            finally {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/diagnosis/OldDataMonitor.java]: NIT: line break before finally

**Comment by oleg-nenashev** [comment]
IMO it impacts job-config-history plugin a bit, because it will be hard to audit changes. It would be useful to get a feedback from @stefanbrausch

Deadlock is worse in any case, so we need to fix the issue somehow.

**Comment by jglick** [comment]
:bee: ([WS-ignoring diff](https://github.com/jenkinsci/jenkins/pull/1796/files?w=1) is easier to review)

**Comment by stephenc** [comment]
@oleg-nenashev I fail to see how this change in any way affects job-config-history. AIUI job-config-history listens independently. I could imagine that a similar bug could affect that plugin, but that would be a separate issue (and would require a different analysis and potentially a different fix)

In any case of job-config-history needs to have its listener run as system it can stash the authentication in a variable (or worst case a `ThreadLocal`) and then go hunting.

:bee:

**Comment by jglick** [comment]
> it impacts job-config-history plugin a bit, because it will be hard to audit changes

I think @oleg-nenashev was referring to that plugin’s need to have a valid `Authentication` associated with changes. But the block being wrapped in this PR is perform read accesses only, not saving anything.

**Comment by oleg-nenashev** [comment]
"removeXYZ" didn't seem to be a read-only block for me. If it is true, then :bee:

**Comment by jtnord** [comment]
@oleg-nenashev `odm.data.remove(referTo(obj))` is `HashMap.remove(Object)` which does not invoke a `save` (which would probably cause an infinite loop if it did!)

**Comment by jtnord** [comment]
@reviewbybees done

**Pull request merged**